pkg/bindings/images/build.go

```diff
@@ -206,46 +226,226 @@ func nTar(sources ...string) (io.ReadCloser, error) {
+					}
+				}
+			}
+
+			err = filepath.Walk(s, func(path string, info os.FileInfo, err error) error {
 				if err != nil {
 					return err
 				}
-				if !info.Mode().IsRegular() || path == s {
-					return nil
-				}
 
-				f, lerr := os.Open(path)
-				if lerr != nil {
-					return lerr
+				if path == s {
+					return nil // skip root dir
 				}
 
 				name := strings.TrimPrefix(path, s+string(filepath.Separator))
-				hdr, lerr := tar.FileInfoHeader(info, name)
-				if lerr != nil {
-					f.Close()
-					return lerr
-				}
-				hdr.Name = name
-				if lerr := tw.WriteHeader(hdr); lerr != nil {
-					f.Close()
-					return lerr
+
+				if !dontApplyExcludes {
+					for _, exclude := range excludes { // ignore docker ignores
+						if m, err := regexpMatch(exclude, name); err != nil {
+							logrus.Errorf("error occurred while regex match: %v", err)
+							return err
+						} else if m {
+							logrus.Debugf("ignoring %v because of rule %v", name, exclude)
+							return nil
+						}
+					}
 				}
 
-				_, cerr := io.Copy(tw, f)
-				f.Close()
-				return cerr
+				if info.Mode().IsRegular() { // add file item
+					f, lerr := os.Open(path)
+					if lerr != nil {
+						return lerr
+					}
+
+					hdr, lerr := tar.FileInfoHeader(info, name)
+					if lerr != nil {
+						f.Close()
+						return lerr
+					}
+					hdr.Name = name
+					if lerr := tw.WriteHeader(hdr); lerr != nil {
+						f.Close()
+						return lerr
+					}
+
+					_, cerr := io.Copy(tw, f)
+					f.Close()
+					return cerr
+				} else if info.Mode().IsDir() { // add folders
+					hdr, lerr := tar.FileInfoHeader(info, name)
+					if lerr != nil {
+						return lerr
+					}
+					hdr.Name = name
+					if lerr := tw.WriteHeader(hdr); lerr != nil {
+						return lerr
+					}
+				} else if info.Mode()&os.ModeSymlink != 0 { // add symlinks as it, not content
+					link, err := os.Readlink(path)
+					if err != nil {
+						return err
+					}
+					hdr, lerr := tar.FileInfoHeader(info, link)
+					if lerr != nil {
+						return lerr
+					}
+					hdr.Name = name
+					if lerr := tw.WriteHeader(hdr); lerr != nil {
+						return lerr
+					}
+				} //skip other than file,folder and symlinks
+				return nil
 			})
 			merr = multierror.Append(merr, err)
 		}
 	}()
 	return pr, merr
 }
+
+func parseIgnoreFile(root string) ([]string, error) {
+	// try with .containerignore
+	cignore, err := ioutil.ReadFile(filepath.Join(root, ".containerignore"))
+	if err != nil && !os.IsNotExist(err) {
+		return nil, fmt.Errorf("error reading .containerignore: '%s'", err)
+	}
+	if err == nil {
+		logrus.Debugf(".containerignore found using it")
+	}
+	// try with .dockerignore
+	logrus.Debugf(".containerignore not found try .dockerignore")
+	dignore, err := ioutil.ReadFile(filepath.Join(root, ".dockerignore"))
+	if err != nil && !os.IsNotExist(err) {
+		return nil, fmt.Errorf("error reading .dockerignore: '%s'", err)
+	}
+	if err == nil {
+		logrus.Debugf(".dockerignore found using it")
+	}
```

Comment:
The code here does not describe the judgment of the situation when both files exist and the description of related documents

Reply:
ok, wont use this approach. from documentation I found that there is command line argument **ignorefile** https://github.com/containers/podman/blob/master/docs/source/markdown/podman-build.1.md#--ignorefile 

however, I cannot found its reference at BuildOptions. Do you have any idea?

Reply:
```
./bin/podman build --help |& grep -- --ignore
      --ignorefile string                            path to an alternate .dockerignore file
```
could see it at the latest code

Reply:
Ok, let me explain, the problem is **--ignorefile** is **buildah**'s internal option as used here: https://github.com/containers/buildah/blob/59cd527d6d40b5a77d7f3e6560b90c2ef7102c93/cmd/buildah/bud.go#L357-L363 buildah reads ignorefile and sets its lines to **options.Excludes**.  **podman** appends buildah's bud args to its build args converts some of them to **imagebuildah.BuildOptions**. And now --ignorefile never applied into Excludes. I put code to check Excludes if empty. If empty than retry read .dockerignore at: https://github.com/containers/podman/pull/8456/files#diff-43fd97ffbb9420a1f310fa0f6a07e2c6ea1d4deb0eefc76ae13812102c11c65fR144

Reply:
SGTM